File: _archive_ts/packages/core/src/governance/authority.ts
/**
 * 能力签发服务（OpenHarness 二开版）
 * 作用：把执行权限从提示词中剥离，显式签发可审计的最小权限令牌。
 */

import { generateId, now } from '@velaris/shared';
import type { AuthorityPlan, CapabilityToken, GovernanceDemand } from '../types.js';

/** 能力签发配置。 */
export interface AuthorityConfig {
  /** 默认令牌 TTL（秒）。 */
  defaultTokenTtlSeconds: number;
  /** 在 ask 模式下触发审批的敏感能力集合。 */
  approvalSensitiveCapabilities: string[];
}

const DEFAULT_AUTHORITY_CONFIG: AuthorityConfig = {
  defaultTokenTtlSeconds: 1800,
  approvalSensitiveCapabilities: ['write', 'exec', 'audit'],
};

/**
 * 授权服务：生成审批判定和能力令牌。
 */
export class AuthorityService {
  private readonly config: AuthorityConfig;

  constructor(config?: Partial<AuthorityConfig>) {
    this.config = {
      ...DEFAULT_AUTHORITY_CONFIG,
      ...config,
    };
  }

  /**
   * 生成授权计划。
   * 为什么这样做：执行前统一落地权限需求，避免隐式越权。
   */
  issuePlan(
    requiredCapabilities: string[],
    governance: GovernanceDemand,
  ): AuthorityPlan {
    const dedupedCapabilities = [...new Set(requiredCapabilities)];
    const approvalsRequired = this.shouldRequireApproval(dedupedCapabilities, governance);
    const capabilityTokens = this.createTokens(dedupedCapabilities);

    return {
      approvalsRequired,
      requiredCapabilities: dedupedCapabilities,
      capabilityTokens,
    };
  }

  /**
   * 计算是否需要审批。
   * 为什么这样做：把审批逻辑收敛到单函数，便于根据合规要求快速调整。
   */
  private shouldRequireApproval(
    requiredCapabilities: string[],
    governance: GovernanceDemand,
  ): boolean {
    if (governance.approvalMode === 'strict') {
      return true;
    }
    if (governance.approvalMode === 'none') {
      return false;
    }
    return requiredCapabilities.some((capability) =>
      this.config.approvalSensitiveCapabilities.includes(capability),
    );
  }

  /**
   * 创建能力令牌。
   * 为什么这样做：每次执行都签发短时令牌，支持最小权限和可追踪失效。
   */
  private createTokens(requiredCapabilities: string[]): CapabilityToken[] {
    if (requiredCapabilities.length === 0) {
      return [];
    }
    return [
      {
        tokenId: generateId('cap'),
        scope: requiredCapabilities,
        ttlSeconds: this.config.defaultTokenTtlSeconds,
        issuedAt: now(),
      },
    ];
  }
}

File: src/velaris_agent/velaris/authority.py
"""Velaris 原生能力签发服务。"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import timezone, datetime
from typing import Any
from uuid import uuid4


@dataclass(frozen=True)
class CapabilityToken:
    """能力令牌。"""

    token_id: str
    scope: list[str]
    ttl_seconds: int
    issued_at: str


@dataclass(frozen=True)
class AuthorityPlan:
    """授权计划。"""

    approvals_required: bool
    required_capabilities: list[str]
    capability_tokens: list[CapabilityToken]

    def to_dict(self) -> dict[str, Any]:
        """把授权计划转换为 JSON 友好的字典。"""
        return asdict(self)


class AuthorityService:
    """能力签发服务。"""

    def __init__(
        self,
        default_token_ttl_seconds: int = 1800,
        approval_sensitive_capabilities: list[str] | None = None,
    ) -> None:
        """初始化授权服务。"""
        self.default_token_ttl_seconds = default_token_ttl_seconds
        self.approval_sensitive_capabilities = approval_sensitive_capabilities or [
            "write",
            "exec",
            "audit",
            "contract_form",
        ]

    def issue_plan(self, required_capabilities: list[str], governance: dict[str, Any]) -> AuthorityPlan:
        """生成授权计划。"""
        deduped_capabilities = list(dict.fromkeys(required_capabilities))
        approval_mode = str(governance.get("approval_mode", "default"))
        approvals_required = approval_mode == "strict" or any(
            capability in self.approval_sensitive_capabilities
            for capability in deduped_capabilities
        )
        tokens = self._create_tokens(deduped_capabilities)
        return AuthorityPlan(
            approvals_required=approvals_required,
            required_capabilities=deduped_capabilities,
            capability_tokens=tokens,
        )

    def _create_tokens(self, required_capabilities: list[str]) -> list[CapabilityToken]:
        """创建短时能力令牌。"""
        if not required_capabilities:
            return []
        return [
            CapabilityToken(
                token_id=f"cap-{uuid4().hex[:12]}",
                scope=required_capabilities,
                ttl_seconds=self.default_token_ttl_seconds,
                issued_at=datetime.now(timezone.utc).isoformat(),
            )
        ]
